### PR TITLE
refactor(Modal): refactor closeIcon

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [X] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- https://github.com/ant-design/ant-design/discussions/42828

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Modal support hidden close button when closeIcon setting to null or false |
| 🇨🇳 Chinese | 弹窗组件支持通过设置 closeIcon 为 null 或 false 隐藏关闭按钮 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 4aa5329</samp>

This pull request adds a new feature to the `Modal` component, which allows the user to hide the close button by setting the `closeIcon` prop to `null` or `false`. It also updates the TypeScript definitions, the test cases, and the documentation for this prop.

### <samp>🤖 Generated by Copilot at 4aa5329</samp>

*  Add feature to hide close button in `Modal` component by setting `closeIcon` prop to `null` or `false` ([link](https://github.com/ant-design/ant-design/pull/43017/files?diff=unified&w=0#diff-a3edcd43b644d0539a67f65e68ba30d6acb77a12531a6a71d1e26a008a9d752cR67), [link](https://github.com/ant-design/ant-design/pull/43017/files?diff=unified&w=0#diff-a3edcd43b644d0539a67f65e68ba30d6acb77a12531a6a71d1e26a008a9d752cR95-R101), [link](https://github.com/ant-design/ant-design/pull/43017/files?diff=unified&w=0#diff-a3edcd43b644d0539a67f65e68ba30d6acb77a12531a6a71d1e26a008a9d752cL108-R117), [link](https://github.com/ant-design/ant-design/pull/43017/files?diff=unified&w=0#diff-ac9fa8e4784b40c5c2a26fe86dc50bbd39572b8ecf42b73611745ec817d883a1L53-R53), [link](https://github.com/ant-design/ant-design/pull/43017/files?diff=unified&w=0#diff-5b32c615603a02b35bc7b0b1651031eb23c5af3d939495dc127088d2efcd58b0R36-R42))
*  Update API documentation for `closeIcon` prop in English and Chinese versions ([link](https://github.com/ant-design/ant-design/pull/43017/files?diff=unified&w=0#diff-0868ee605478aef1b7a7e4508dce443964207f1dd966d82be9bf7e9b16a31aa5L50-R50), [link](https://github.com/ant-design/ant-design/pull/43017/files?diff=unified&w=0#diff-7a204621cc9425007680c097ad01151df377979692bbc09eb2796025b85043f3L51-R51))
*  Add comment to deprecate `closable` prop and recommend using `closeIcon` prop instead ([link](https://github.com/ant-design/ant-design/pull/43017/files?diff=unified&w=0#diff-ac9fa8e4784b40c5c2a26fe86dc50bbd39572b8ecf42b73611745ec817d883a1L11-R11))

### CODE REVIEW

File: components/modal/Modal.tsx [hunk 1]
```diff
@@ -91,6 +92,13 @@ const Modal: React.FC<ModalProps> = (props) => {
   const dialogFooter =
     footer === undefined ? <Footer {...props} onOk={handleOk} onCancel={handleCancel} /> : footer;
 
+  const mergedClosable = React.useMemo(() => {
```
Comment (zombieJ) [inline components/modal/Modal.tsx:95]: 这个基本都是这个套路，直接抽一个 useCloable hooks 吧，把 closable 和 closeIcon 通一提供出来

File: components/drawer/DrawerPanel.tsx [hunk 2]
```diff
@@ -44,13 +45,7 @@ const DrawerPanel: React.FC<DrawerPanelProps> = (props) => {
...
   } = props;
 
-  const mergedClosable = React.useMemo(() => {
-    if (typeof closable === 'boolean') {
-      return closable;
-    }
-
-    return closeIcon !== null && closeIcon !== false;
-  }, [closable, closeIcon]);
+  const mergedClosable = useClosable(closable, closeIcon, true);
```
Comment (zombieJ) [inline components/drawer/DrawerPanel.tsx:48]: 下面 mergedCloseIcon 没吃进去

File: components/modal/Modal.tsx [hunk 3]
```diff
@@ -105,7 +109,8 @@ const Modal: React.FC<ModalProps> = (props) => {
           visible={open ?? visible}
           mousePosition={restProps.mousePosition ?? mousePosition}
           onClose={handleCancel}
-          closeIcon={renderCloseIcon(prefixCls, closeIcon)}
+          closable={mergedClosable}
+          closeIcon={mergedClosable ? renderCloseIcon(prefixCls, closeIcon) : null}
```
Comment (zombieJ) [inline components/modal/Modal.tsx:113]: 同上，其他的都看一下。应该都是 `useClosable` 来集中处理的

File: components/_util/hooks/useClosable.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { CloseOutlined } from '@ant-design/icons';
+import type { ReactNode } from 'react';
+import React from 'react';
+
+function useInnerClosable(
```
Comment (zombieJ) [inline components/_util/hooks/useClosable.tsx:5]: 看起啦没有必要拆成 2 个 hooks

File: components/_util/hooks/useClosable.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+export type UseClosableParams = {
+  closable?: boolean;
+  closeIcon?: boolean | ReactNode;
+  defaultClosable?: boolean;
+  defaultCloseIcon?: ReactNode;
+  customCloseIconRender?: (closeIcon: ReactNode) => ReactNode;
+};
+
+export default function useClosable({
```
Comment (zombieJ) [inline components/_util/hooks/useClosable.tsx:27]: 这里还是平铺 param 比较好，obj 无法被 uglify 掉

File: components/_util/hooks/useClosable.tsx (hunk 4 above)
Comment (kiner-tang) [reply components/_util/hooks/useClosable.tsx:5]: 我们要先计算出是否需要展示 closeIcon , 才能确定 应该返回 怎样的 closeIcon , 所以这边把逻辑抽离了一下

File: components/_util/hooks/useClosable.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  customCloseIconRender?: (closeIcon: ReactNode) => ReactNode;
+};
+
+export default function useClosable(
+  closable?: boolean,
+  closeIcon?: boolean | ReactNode,
+  customCloseIconRender?: (closeIcon: ReactNode) => ReactNode,
+  defaultCloseIcon: ReactNode = <CloseOutlined />,
+  defaultClosable = false,
+): [boolean, React.ReactNode | null] {
```
Comment (zombieJ) [inline components/_util/hooks/useClosable.tsx:33]:
```diff
-- ): [boolean, React.ReactNode | null] {
++ ): [closable: boolean, closeIcon: React.ReactNode | null] {
```

File: components/_util/hooks/useClosable.tsx [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  const mergedClosable = useInnerClosable(closable, closeIcon, defaultClosable);
+  if (!mergedClosable || closeIcon === null || closeIcon === false) {
```
Comment (zombieJ) [inline components/_util/hooks/useClosable.tsx:35]: 上面 `mergedClosable` 判断出来后，这边不应该再判断 `closeIcon`。

File: components/modal/index.en-US.md [hunk 8]
```diff
@@ -47,8 +47,7 @@ Additionally, if you need show a simple confirmation dialog, you can use [`App.u
 | cancelButtonProps | The cancel button props | [ButtonProps](/components/button/#api) | - |  |
 | cancelText | Text of the Cancel button | ReactNode | `Cancel` |  |
 | centered | Centered Modal | boolean | false |  |
-| closable | Whether a close (x) button is visible on top right of the modal dialog or not | boolean | true |  |
-| closeIcon | Custom close icon | ReactNode | &lt;CloseOutlined /> |  |
+| closeIcon | Custom close icon. 5.7.0: close button will be hidden when setting to null or false | boolean \| ReactNode | &lt;CloseOutlined /> |  |
```
Comment (zombieJ) [inline components/modal/index.en-US.md:50]: ``` `null` ``` 反引号又没啦~

Comment (kiner-tang) [reply components/modal/index.en-US.md:50]: 不对呀，这个昨天补过的，咋又丢了？

**Review by zombieJ — APPROVED** [review]

File: components/modal/index.en-US.md (hunk 8 above)
Comment (kiner-tang) [reply components/modal/index.en-US.md]: done

**Pull request merged**